src/Servers/Kestrel/Core/src/Internal/Http/Http1MessageBody.cs

```diff
@@ -299,14 +299,87 @@ public ForUpgrade(Http1Connection context)
                 RequestUpgrade = true;
             }
 
+            // This returns IsEmpty so we can avoid draining the body (since it's basically an endless stream)
             public override bool IsEmpty => true;
 
-            protected override bool Read(ReadOnlySequence<byte> readableBuffer, PipeWriter writableBuffer, out SequencePosition consumed, out SequencePosition examined)
+            public override async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+            {
+                while (true)
+                {
+                    var result = await _context.Input.ReadAsync(cancellationToken);
+                    var readableBuffer = result.Buffer;
+                    var readableBufferLength = readableBuffer.Length;
+
+                    try
+                    {
+                        if (!readableBuffer.IsEmpty)
+                        {
+                            foreach (var memory in readableBuffer)
+                            {
+                                // REVIEW: This *could* be slower if 2 things are true
+                                // - The WriteAsync(ReadOnlyMemory<byte>) isn't overridden on the destination
+                                // - We change the Kestrel Memory Pool to not use pinned arrays but instead use native memory
+                                await destination.WriteAsync(memory, cancellationToken);
+                            }
+                        }
+
+                        if (result.IsCompleted)
+                        {
+                            return;
+                        }
+                    }
+                    finally
+                    {
+                        _context.Input.AdvanceTo(readableBuffer.End);
+                    }
+                }
+            }
+
+            public override async ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken = default)
+            {
+                while (true)
+                {
+                    var result = await _context.Input.ReadAsync(cancellationToken);
+                    var readableBuffer = result.Buffer;
+                    var readableBufferLength = readableBuffer.Length;
+
+                    var consumed = readableBuffer.End;
+                    var actual = 0;
+
+                    try
+                    {
+                        if (!readableBuffer.IsEmpty)
+                        {
+                            // buffer.Length is int
+                            actual = (int)Math.Min(readableBufferLength, buffer.Length);
+
+                            var slice = readableBuffer.Slice(0, actual);
+                            consumed = readableBuffer.GetPosition(actual);
+                            slice.CopyTo(buffer.Span);
```

Comment:
While you're at it, you might as well make the changes I recommended in https://github.com/aspnet/AspNetCore/pull/6583#discussion_r247287427 here and in MessageBody.

Reply:
@jkotalik Already made the MessageBody change https://github.com/aspnet/AspNetCore/pull/6583/commits/dd4fd887180ac493fda2ff07b21798429b084288. Just change it here to match.